---
title: sQTLseekeR Overview
author: Jean Monlong
date: October 5, 2016
output:
   md_document:
     fig_width: 10
     variant: markdown_github
---

# sQTLseekeR Overview

```{r, include=FALSE}
knitr::opts_chunk$set(message=FALSE, warning=FALSE)
```

```{r}
library(sQTLseekeR)
```

## Preparing the data

In the [Data](https://github.com/guigolab/sQTLseekeR/tree/master/Data) folder, you can find the transcript expressions for 5 genes and SNPs in these 5 regions.
The genes coordinates are present in a third file.

```{r}
trans.exp.f = "../Data/transExpression.tsv.gz"
gene.bed.f = "../Data/genes.bed"
genotype.f = "../Data/snps-012coded.tsv"
```

First, the **ordered** genotype file should be compressed and indexed (if not done externally before):

```{r}
genotype.indexed.f = index.genotype(genotype.f)
```

Then the transcript expression should be imported, cleaned (removing genes with low expression, only one transcripts, etc).
`prepare.trans.exp` returns relative transcript expressions.

```{r}
te.df = read.table(trans.exp.f, as.is=TRUE, header=TRUE, sep="\t")
tre.df = prepare.trans.exp(te.df)
tre.df[1:5,1:5]
```

## Test for gene/SNP associations

To run the tests, simply feed the transcript relative expression, genotype file path and the genes coordinates:

```{r}
gene.bed = read.table(gene.bed.f, as.is=TRUE, sep="\t")
colnames(gene.bed) = c("chr","start","end","geneId")
res.df = sqtl.seeker(tre.df, genotype.indexed.f, gene.bed, svQTL=TRUE, verbose=FALSE)
head(res.df)
```

The resulting data.frame shows information about the association between each SNP/gene.

In practice it's good to save these results before proceeding to sQTLs selection.

```{r}
write.table(res.df, file="sQTLs-all.tsv", quote=FALSE, row.names=FALSE, sep="\t")
```

Finally, sQTLs are selected after multiple-testing correction with a user-defined FDR threshold.

```{r}
sqtls.df = sqtls(res.df, FDR=.01, out.pdf="sQTLs-FDR01.pdf")
head(sqtls.df)
```


## Characterizing one sQTL

Let's say we want to visualize the effect of one of the detected sQTL.
The first thing to look at is the relative expression of each transcript in the different genotype groups.

I also want to know the structure of each transcript so I load the CDS/UTR structure.
This information was formatted from Genecode v12 annotation (see the [Splicing Event Classification tutorial](SplicingEventClassification.md) for more details).

```{r}
load("../Data/transcriptStructureGencodeV12.RData")
sqtl.plot(sqtls.df$geneId[1], sqtls.df$snpId[1], tre.df, genotype.f, gene.bed, tr.str=trans.str)
```

With these graphs, we see that it's mostly the first and second transcript whose usage changes, increasing and decreasing respectively.
The relative expression of the other transcripts mostly stays the same across the different genotype groups.

The schematic structure of the transcripts show that these two transcripts differ quite a lot, and both at the beginning and end of the gene body.

## R session

```{r}
sessionInfo()
```
